tools/nl/embeddings/file_util.py

```diff
@@ -0,0 +1,97 @@
+import io
+import os
+
+from google.cloud import storage
+
+_GCS_PATH_PREFIX = "gs://"
+
+
+class FileHandler:
+  """(Abstract) base class that should be extended by concrete implementations."""
+
+  def __init__(self, path: str) -> None:
+    self.path = path
+
+  def __str__(self) -> str:
+    return self.path
+
+  def read_string(self) -> str:
+    pass
+
+  def read_string_io(self) -> io.StringIO:
+    return io.StringIO(self.read_string())
+
+  def write_string(self, content: str) -> None:
+    pass
+
+  def join(self, subpath: str) -> str:
+    return os.path.join(self.path, subpath)
+
+  def abspath(self) -> str:
+    return self.path
+
+
+class LocalFileHandler(FileHandler):
+
+  def __init__(self, path: str) -> None:
+    super().__init__(path)
+
+  def read_string(self) -> str:
+    with open(self.path, "r") as f:
+      return f.read()
+
+  def write_string(self, content: str) -> None:
+    with open(self.path, "w") as f:
+      f.write(content)
+
+  def abspath(self) -> str:
+    return os.path.abspath(self.path)
+
+
+class GcsMeta(type):
+
+  @property
+  def gcs_client(cls) -> storage.Client:
+    """Creates GCS client as a lazy class property.
+    Class property ensures that only 1 client instance is created across all GcsFileHandler instances.
+    Lazy ensures that a client is not created if no GcsFileHandler instances are created.
```

Comment:
interesting.  where is the `lazy` part specified?

Reply:
This method is called whenever `GcsFileHandler.gcs_client` is accessed. The implementation creates it lazily and caches it. See: https://stackoverflow.com/a/15226813